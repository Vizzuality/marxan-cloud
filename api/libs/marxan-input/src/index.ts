export { MarxanFileInput } from './marxan-file-input';
export {
  RunMode,
  ClumpType,
  HeuristicType,
  IterativeImprovementType,
  MarxanParameters,
} from './marxan-parameters';
export { MarxanInput } from './marxan-input';
