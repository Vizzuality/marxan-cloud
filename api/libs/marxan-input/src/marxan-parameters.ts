import { IsEnum, IsNumber, IsOptional } from 'class-validator';

export enum RunMode {
  SimulatedAnnealingFollowedByHeuristic = 0,
  SimulatedAnnealingFollowedByIterativeImprovement = 1,
  SimulatedAnnealingFollowedByHeuristicAndIterativeImprovement = 2,
  OnlyHeuristic = 3,
  OnlyIterativeImprovement = 4,
  HeuristicAndIterativeImprovement = 5,
  OnlySimulatedAnnealing = 6,
}

export enum IterativeImprovementType {
  Normal = 0,
  TwoStep = 1,
  Swap = 2,
  NormalFollowedByTwoStep = 3,
}

export enum HeuristicType {
  Ignored = -1,
  Richness = 0,
  Greedy = 1,
  MaxRarity = 2,
  BestRarity = 3,
  AverageRarity = 4,
  SumRarity = 5,
  ProductIrreplaceability = 6,
  SummationIrreplaceability = 7,
}

export enum ClumpType {
  Ignored = -1,
  PartialDoNotCount = 0,
  PartialCountAsHalf = 1,
  GraduatedPenalty = 3,
}

export class MarxanParametersDefaults {
  BLM = 1;
  PROP = 0.5;
  RANDSEED = -1;
  BESTSCORE = 0;
  NUMREPS = 10;
  NUMITNS = 1000000;
  STARTTEMP = -1;
  COOLFAC = 0;
  NUMTEMP = 10000;
  COSTTHRESH = 0;
  THRESHPEN1 = 0;
  THRESHPEN2 = 0;
  RUNMODE = RunMode.SimulatedAnnealingFollowedByIterativeImprovement;
  MISSLEVEL = 1;
  ITIMPTYPE = IterativeImprovementType.Normal;
  HEURTYPE = HeuristicType.Ignored;
  CLUMPTYPE = ClumpType.PartialDoNotCount;
  _CLOUD_PROJECT = 'n/a';
  _CLOUD_ORGANIZATION = 'n/a';
  VERBOSITY = 2;
  SCENNAME = 'output';
  SAVESOLUTIONSMATRIX = 3;
  SAVERUN = 3;
  SAVEBEST = 3;
  SAVESUMMARY = 3;
  SAVESCEN = 3;
  SAVETARGMET = 3;
  SAVESUMSOLN = 3;
  SAVELOG = 3;
  SAVESNAPSTEPS = 0;
  SAVESNAPCHANGES = 0;
  SAVESNAPFREQUENCY = 0;
}

const marxanDefaults = new MarxanParametersDefaults();

export class MarxanParameters {
  /**
   * Boundary Length Modifier
   */
  @IsOptional()
  @IsNumber()
  BLM?: number = marxanDefaults.BLM;

  /**
   * Proportion of planning units ininitial reserve system
   */
  @IsOptional()
  @IsNumber()
  PROP?: number = marxanDefaults.PROP;

  /**
   * Random seed number
   */
  @IsOptional()
  @IsNumber()
  RANDSEED?: number = marxanDefaults.RANDSEED;

  /**
   * Best score hint
   */
  @IsOptional()
  @IsNumber()
  BESTSCORE?: number = marxanDefaults.BESTSCORE;

  /**
   * Number of repeat runs (or solutions)
   */
  @IsOptional()
  @IsNumber()
  NUMREPS?: number = marxanDefaults.NUMREPS;

  // Annealing

  /**
   * Number of iterations for annealing
   */
  @IsOptional()
  @IsNumber()
  NUMITNS?: number = marxanDefaults.NUMITNS;

  /**
   * Number of iterations for annealing
   */
  @IsOptional()
  @IsNumber()
  STARTTEMP?: number = marxanDefaults.STARTTEMP;

  /**
   * Cooling factor for annealing
   */
  @IsOptional()
  @IsNumber()
  COOLFAC?: number = marxanDefaults.COOLFAC;

  /**
   * Number of temperature decreases for annealing
   */
  @IsOptional()
  @IsNumber()
  NUMTEMP?: number = marxanDefaults.NUMTEMP;

  // Cost Threshold

  /**
   * Cost threshold
   */
  @IsOptional()
  @IsNumber()
  COSTTHRESH?: number = marxanDefaults.COSTTHRESH;

  /**
   * Size of cost threshold penalty
   */
  @IsOptional()
  @IsNumber()
  THRESHPEN1?: number = marxanDefaults.THRESHPEN1;

  /**
   * Shape of cost threshold penalty
   */
  @IsOptional()
  @IsNumber()
  THRESHPEN2?: number = marxanDefaults.THRESHPEN2;

  // Program Control

  @IsOptional()
  @IsEnum(RunMode)
  RUNMODE?: RunMode = marxanDefaults.RUNMODE;

  /**
   * Species missing proportion
   */
  @IsOptional()
  MISSLEVEL?: number = marxanDefaults.MISSLEVEL;

  @IsOptional()
  @IsEnum(IterativeImprovementType)
  ITIMPTYPE?: IterativeImprovementType = marxanDefaults.ITIMPTYPE;

  @IsOptional()
  @IsEnum(HeuristicType)
  HEURTYPE?: HeuristicType = marxanDefaults.HEURTYPE;

  @IsOptional()
  @IsEnum(ClumpType)
  CLUMPTYPE?: ClumpType = marxanDefaults.CLUMPTYPE;
}
