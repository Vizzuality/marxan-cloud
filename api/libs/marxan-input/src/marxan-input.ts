import { Injectable } from '@nestjs/common';
import { plainToClass } from 'class-transformer';
import { validateSync } from 'class-validator';
import { MarxanParameters } from './marxan-parameters';

@Injectable()
export class MarxanInput {
  from(input: Partial<MarxanParameters>) {
    const params = plainToClass<MarxanParameters, MarxanParameters>(
      MarxanParameters,
      input,
    );
    const errors = validateSync(params, {
      whitelist: true,
    });

    if (errors.length > 0) {
      throw errors;
    }

    return params;
  }
}
