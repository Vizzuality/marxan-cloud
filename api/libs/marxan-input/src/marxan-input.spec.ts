import { MarxanInput } from './marxan-input';
import { HeuristicType, RunMode } from './marxan-parameters';

const sut = new MarxanInput();

describe(`when providing unsupported values for enums`, () => {
  it(`should throw`, () => {
    expect.assertions(1);
    try {
      sut.from({
        HEURTYPE: 9999999 as HeuristicType,
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe(`when unwanted properties are provided`, () => {
  it(`should skip them`, () => {
    const params = sut.from({
      VERBOSITY: 3,
      SAVESUMMARY: 0,
      OUTPUTDIR: '/etc/passwd',
    } as any);

    expect('VERBOSITY' in params).toBeFalsy();
    expect('SAVESUMMARY' in params).toBeFalsy();
    expect('OUTPUTDIR' in params).toBeFalsy();
    expect('BLM' in params).toBeTruthy();
  });
});

describe(`when providing some values`, () => {
  const input = {
    HEURTYPE: HeuristicType.SummationIrreplaceability,
    MISSLEVEL: 999,
    BESTSCORE: 1337,
    RUNMODE: RunMode.OnlySimulatedAnnealing,
  };
  it(`should override defaults`, () => {
    const params = sut.from(input);
    expect(params).toMatchSnapshot();
  });
});

describe(`when empty input is provided`, () => {
  it(`should resolve to defaults`, () => {
    expect(sut.from({})).toMatchInlineSnapshot(`
      MarxanParameters {
        "BESTSCORE": 0,
        "BLM": 1,
        "CLUMPTYPE": 0,
        "COOLFAC": 0,
        "COSTTHRESH": 0,
        "HEURTYPE": -1,
        "ITIMPTYPE": 0,
        "MISSLEVEL": 1,
        "NUMITNS": 1000000,
        "NUMREPS": 10,
        "NUMTEMP": 10000,
        "PROP": 0.5,
        "RANDSEED": -1,
        "RUNMODE": 1,
        "STARTTEMP": -1,
        "THRESHPEN1": 0,
        "THRESHPEN2": 0,
      }
    `);
  });
});
