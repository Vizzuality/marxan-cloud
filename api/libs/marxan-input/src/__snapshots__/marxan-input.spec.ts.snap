// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when providing some values should override defaults 1`] = `
MarxanParameters {
  "BESTSCORE": 1337,
  "BLM": 1,
  "CLUMPTYPE": 0,
  "COOLFAC": 0,
  "COSTTHRESH": 0,
  "HEURTYPE": 7,
  "ITIMPTYPE": 0,
  "MISSLEVEL": 999,
  "NUMITNS": 1000000,
  "NUMREPS": 10,
  "NUMTEMP": 10000,
  "PROP": 0.5,
  "RANDSEED": -1,
  "RUNMODE": 6,
  "STARTTEMP": -1,
  "THRESHPEN1": 0,
  "THRESHPEN2": 0,
}
`;

exports[`when providing unsupported values for enums should throw 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isEnum": "HEURTYPE must be a valid enum value",
    },
    "property": "HEURTYPE",
    "target": MarxanParameters {
      "BESTSCORE": 0,
      "BLM": 1,
      "CLUMPTYPE": 0,
      "COOLFAC": 0,
      "COSTTHRESH": 0,
      "HEURTYPE": 9999999,
      "ITIMPTYPE": 0,
      "MISSLEVEL": 1,
      "NUMITNS": 1000000,
      "NUMREPS": 10,
      "NUMTEMP": 10000,
      "PROP": 0.5,
      "RANDSEED": -1,
      "RUNMODE": 1,
      "STARTTEMP": -1,
      "THRESHPEN1": 0,
      "THRESHPEN2": 0,
    },
    "value": 9999999,
  },
]
`;
